oak_functions_service/src/lookup_htbl.rs

```diff
@@ -0,0 +1,575 @@
+//
+// Copyright 2024 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// This is a key/value hash table optimized for large numbers of small k/v pairs.  We assume we
+// bust out of on-chip cache, so speed is limited by cache misses, not the actual operations such
+// as hashing and bit manipulation.
+//
+// The k/v pairs are loaded before any are read, and once they are loaded cannot be modified.  We
+// take advantage of these restrictions to build a hash table with far less memory overhead per k/v
+// pair (21 bytes vs ~128 bytes for hashbrown<Bytes, Bytes>).  Speed for k/v lookups is ~90ns on
+// manual tests.
+
+use alloc::{boxed::Box, vec, vec::Vec};
+use core::mem;
+
+use bytes::Bytes;
+
+// To save memory, we use 5 byte array index values instead of 8 bytes.  This saves 12 bytes per
+// k/v pair.  It is referred to as u40 below.
+const INDEX_SIZE: usize = 5;
+// Size of an entry in the hash table.
+const ENTRY_SIZE: usize = INDEX_SIZE + 1;
+
+// The hash table consists of 2 vectors.  The first is a table of "entries", which is a packed
+// array of entries with the following layout:
```

Comment:
Could this be implemented as an actual proper data structure?

Something like this:
```
#[repr(C, packed)]
struct Entry {
  hash_byte: u8,
  data_index: [u8; 5],
}
```

and the Vec could be of those structs, not raw [u8].
Alternatively, if the store is still a bag of bytes, the `zerocopy` crate might be useful when transforming these structs into raw bytes and back.

Reply:
Done.